File: backend/app/models/promotion.py
"""Promotion domain model for discounts and coupons.

This module defines the Promotion entity for managing promotional codes,
discounts, and coupon redemptions. Supports percentage discounts, fixed
amounts, and trial extensions with flexible eligibility rules stored as JSONB.
"""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class DiscountType(str, Enum):
    """Type of discount applied.

    Defines how the discount_value should be interpreted.
    """

    PERCENTAGE = "percentage"
    FIXED_AMOUNT = "fixed_amount"
    FREE_TRIAL_EXTENSION = "free_trial_extension"


class PromotionStatus(str, Enum):
    """Current status of a promotion.

    Tracks the lifecycle state of a promotional code.
    """

    ACTIVE = "active"
    EXPIRED = "expired"
    EXHAUSTED = "exhausted"
    DISABLED = "disabled"


class PromotionBase(BaseModel):
    """Base promotion fields shared across DTOs.

    Contains core promotion configuration fields.
    """

    code: str = Field(
        ...,
        min_length=3,
        max_length=50,
        description="Unique promotion code (auto-uppercased)",
    )
    name: str = Field(
        ...,
        min_length=1,
        max_length=100,
        description="Display name for the promotion",
    )
    description: str | None = Field(
        default=None,
        max_length=500,
        description="Optional description of the promotion",
    )
    discount_type: DiscountType = Field(
        ...,
        description="How the discount is applied",
    )
    discount_value: Decimal = Field(
        ...,
        gt=0,
        description="Discount amount (percentage or fixed)",
    )
    duration_months: int | None = Field(
        default=None,
        ge=1,
        le=36,
        description="How many months the discount applies",
    )
    max_redemptions: int | None = Field(
        default=None,
        ge=1,
        description="Maximum number of redemptions (None = unlimited)",
    )
    valid_from: datetime = Field(
        ...,
        description="When the promotion becomes valid",
    )
    valid_until: datetime | None = Field(
        default=None,
        description="When the promotion expires",
    )
    eligibility_rules: dict[str, Any] = Field(
        default_factory=dict,
        description="JSONB rules for eligibility (e.g., new_customers_only)",
    )

    @field_validator("code")
    @classmethod
    def uppercase_code(cls, v: str) -> str:
        """Convert promotion code to uppercase."""
        return v.upper()

    @model_validator(mode="after")
    def validate_percentage_range(self) -> "PromotionBase":
        """Ensure percentage discounts don't exceed 100%."""
        if self.discount_type == DiscountType.PERCENTAGE:
            if self.discount_value > Decimal("100"):
                raise ValueError("Percentage discount cannot exceed 100%")
        return self


class PromotionCreate(PromotionBase):
    """DTO for creating a new promotion.

    Used when creating promotional codes with optional Stripe integration.
    """

    stripe_coupon_id: str | None = Field(
        default=None,
        description="Stripe coupon ID for integration",
    )


class PromotionUpdate(BaseModel):
    """DTO for updating an existing promotion.

    All fields are optional - only provided fields are updated.
    Note: code and discount_type cannot be changed after creation.
    """

    name: str | None = Field(
        default=None,
        min_length=1,
        max_length=100,
        description="Updated display name",
    )
    description: str | None = Field(
        default=None,
        max_length=500,
        description="Updated description",
    )
    max_redemptions: int | None = Field(
        default=None,
        ge=1,
        description="Updated max redemptions",
    )
    valid_until: datetime | None = Field(
        default=None,
        description="Updated expiration date",
    )
    status: PromotionStatus | None = Field(
        default=None,
        description="Updated promotion status",
    )
    eligibility_rules: dict[str, Any] | None = Field(
        default=None,
        description="Updated eligibility rules",
    )


class Promotion(PromotionBase):
    """Full promotion model with all fields.

    Represents a complete promotion record as stored in the database.
    Includes redemption tracking and Stripe integration fields.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID = Field(
        ...,
        description="Unique promotion identifier",
    )
    current_redemptions: int = Field(
        default=0,
        ge=0,
        description="Number of times this promotion has been redeemed",
    )
    stripe_coupon_id: str | None = Field(
        default=None,
        description="Stripe coupon ID for integration",
    )
    status: PromotionStatus = Field(
        default=PromotionStatus.ACTIVE,
        description="Current promotion status",
    )
    created_at: datetime = Field(
        ...,
        description="When the promotion was created",
    )
    updated_at: datetime = Field(
        ...,
        description="When the promotion was last updated",
    )


class PromotionRedemption(BaseModel):
    """Record of a promotion being redeemed by an organization.

    Tracks when and by whom a promotion was used.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID = Field(
        ...,
        description="Unique redemption identifier",
    )
    promotion_id: UUID = Field(
        ...,
        description="The promotion that was redeemed",
    )
    organization_id: UUID = Field(
        ...,
        description="The organization that redeemed the promotion",
    )
    redeemed_at: datetime = Field(
        ...,
        description="When the redemption occurred",
    )
    stripe_discount_id: str | None = Field(
        default=None,
        description="Stripe discount ID if applicable",
    )


class PromotionSummary(BaseModel):
    """Lightweight promotion view for listings.

    Contains essential promotion info without full details.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID = Field(description="Unique promotion identifier")
    code: str = Field(description="Promotion code")
    name: str = Field(description="Display name")
    discount_type: DiscountType = Field(description="Type of discount")
    discount_value: Decimal = Field(description="Discount amount")
    status: PromotionStatus = Field(description="Current status")
    current_redemptions: int = Field(description="Number of redemptions")
    max_redemptions: int | None = Field(description="Max redemptions allowed")
    valid_until: datetime | None = Field(description="Expiration date")

File: frontend/src/types/promotion.ts
/**
 * Promotion domain types for discounts and coupons.
 *
 * This module defines the Promotion entity for managing promotional codes,
 * discounts, and coupon redemptions. Supports percentage discounts, fixed
 * amounts, and trial extensions with flexible eligibility rules.
 */

import { z } from 'zod'
import { formatMoney } from '@/lib/money'

/**
 * Type of discount applied.
 *
 * Defines how the discount_value should be interpreted.
 */
export const DiscountType = {
  PERCENTAGE: 'percentage',
  FIXED_AMOUNT: 'fixed_amount',
  FREE_TRIAL_EXTENSION: 'free_trial_extension',
} as const
export type DiscountType = (typeof DiscountType)[keyof typeof DiscountType]

/**
 * Current status of a promotion.
 *
 * Tracks the lifecycle state of a promotional code.
 */
export const PromotionStatus = {
  ACTIVE: 'active',
  EXPIRED: 'expired',
  EXHAUSTED: 'exhausted',
  DISABLED: 'disabled',
} as const
export type PromotionStatus =
  (typeof PromotionStatus)[keyof typeof PromotionStatus]

/**
 * Zod schema for discount type validation.
 */
export const DiscountTypeSchema = z.enum([
  'percentage',
  'fixed_amount',
  'free_trial_extension',
])

/**
 * Zod schema for promotion status validation.
 */
export const PromotionStatusSchema = z.enum([
  'active',
  'expired',
  'exhausted',
  'disabled',
])

/**
 * Zod schema for eligibility rules.
 *
 * Flexible schema that allows additional properties for future rule types.
 */
export const EligibilityRulesSchema = z
  .object({
    first_n_users: z.number().optional(),
    plan_restriction: z.array(z.string()).optional(),
    new_customers_only: z.boolean().optional(),
    one_per_organization: z.boolean().optional(),
  })
  .passthrough()

export type EligibilityRules = z.infer<typeof EligibilityRulesSchema>

/**
 * Full promotion schema with all fields.
 */
export const PromotionSchema = z.object({
  id: z.string().uuid(),
  code: z
    .string()
    .min(3)
    .max(50)
    .transform((v) => v.toUpperCase()),
  name: z.string().min(1).max(100),
  description: z.string().max(500).nullable(),
  discount_type: DiscountTypeSchema,
  discount_value: z
    .string()
    .or(z.number())
    .transform((v) => String(v)),
  duration_months: z.number().min(1).max(36).nullable(),
  max_redemptions: z.number().min(1).nullable(),
  current_redemptions: z.number().default(0),
  valid_from: z.string().datetime(),
  valid_until: z.string().datetime().nullable(),
  eligibility_rules: EligibilityRulesSchema.default({}),
  stripe_coupon_id: z.string().nullable(),
  status: PromotionStatusSchema,
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
})

export type Promotion = z.infer<typeof PromotionSchema>

/**
 * Schema for creating a new promotion.
 */
export const PromotionCreateSchema = z.object({
  code: z.string().min(3).max(50),
  name: z.string().min(1).max(100),
  description: z.string().max(500).nullable().optional(),
  discount_type: DiscountTypeSchema,
  discount_value: z.string().or(z.number()),
  duration_months: z.number().min(1).max(36).nullable().optional(),
  max_redemptions: z.number().min(1).nullable().optional(),
  valid_from: z.string().datetime(),
  valid_until: z.string().datetime().nullable().optional(),
  eligibility_rules: EligibilityRulesSchema.optional(),
  stripe_coupon_id: z.string().nullable().optional(),
})

export type PromotionCreate = z.infer<typeof PromotionCreateSchema>

/**
 * Schema for updating an existing promotion.
 */
export const PromotionUpdateSchema = z.object({
  name: z.string().min(1).max(100).optional(),
  description: z.string().max(500).nullable().optional(),
  max_redemptions: z.number().min(1).nullable().optional(),
  valid_until: z.string().datetime().nullable().optional(),
  status: PromotionStatusSchema.optional(),
  eligibility_rules: EligibilityRulesSchema.optional(),
})

export type PromotionUpdate = z.infer<typeof PromotionUpdateSchema>

/**
 * Schema for promotion redemption records.
 */
export const PromotionRedemptionSchema = z.object({
  id: z.string().uuid(),
  promotion_id: z.string().uuid(),
  organization_id: z.string().uuid(),
  redeemed_at: z.string().datetime(),
  stripe_discount_id: z.string().nullable(),
})

export type PromotionRedemption = z.infer<typeof PromotionRedemptionSchema>

/**
 * Lightweight promotion view for listings.
 */
export const PromotionSummarySchema = z.object({
  id: z.string().uuid(),
  code: z.string(),
  name: z.string(),
  discount_type: DiscountTypeSchema,
  discount_value: z
    .string()
    .or(z.number())
    .transform((v) => String(v)),
  status: PromotionStatusSchema,
  current_redemptions: z.number(),
  max_redemptions: z.number().nullable(),
  valid_until: z.string().datetime().nullable(),
})

export type PromotionSummary = z.infer<typeof PromotionSummarySchema>

/**
 * Type guard for valid discount type.
 */
export function isValidDiscountType(value: unknown): value is DiscountType {
  return DiscountTypeSchema.safeParse(value).success
}

/**
 * Type guard for valid promotion status.
 */
export function isValidPromotionStatus(
  value: unknown
): value is PromotionStatus {
  return PromotionStatusSchema.safeParse(value).success
}

/**
 * Get display name for discount type.
 */
export function getDiscountTypeDisplayName(type: DiscountType): string {
  const displayNames: Record<DiscountType, string> = {
    percentage: 'Percentage',
    fixed_amount: 'Fixed Amount',
    free_trial_extension: 'Free Trial Extension',
  }
  return displayNames[type]
}

/**
 * Get display name for promotion status.
 */
export function getPromotionStatusDisplayName(status: PromotionStatus): string {
  const displayNames: Record<PromotionStatus, string> = {
    active: 'Active',
    expired: 'Expired',
    exhausted: 'Exhausted',
    disabled: 'Disabled',
  }
  return displayNames[status]
}

/**
 * Check if promotion is currently usable.
 */
export function isPromotionActive(status: PromotionStatus): boolean {
  return status === 'active'
}

/**
 * Check if promotion has remaining redemptions.
 */
export function hasRemainingRedemptions(
  currentRedemptions: number,
  maxRedemptions: number | null
): boolean {
  if (maxRedemptions === null) {
    return true // Unlimited
  }
  return currentRedemptions < maxRedemptions
}

/**
 * Calculate remaining redemptions.
 */
export function getRemainingRedemptions(
  currentRedemptions: number,
  maxRedemptions: number | null
): number | null {
  if (maxRedemptions === null) {
    return null // Unlimited
  }
  return Math.max(0, maxRedemptions - currentRedemptions)
}

/**
 * Format discount value for display.
 */
export function formatDiscountValue(
  value: string | number,
  type: DiscountType,
  currency: string = 'usd'
): string {
  const numValue = typeof value === 'string' ? parseFloat(value) : value

  switch (type) {
    case 'percentage':
      return `${numValue}%`
    case 'fixed_amount':
      return formatMoney(numValue, currency)
    case 'free_trial_extension':
      return `${numValue} days free`
    default:
      return String(numValue)
  }
}

/**
 * Check if promotion is within valid date range.
 */
export function isPromotionInDateRange(
  validFrom: string,
  validUntil: string | null,
  checkDate: Date = new Date()
): boolean {
  const from = new Date(validFrom)
  if (checkDate < from) {
    return false
  }
  if (validUntil !== null) {
    const until = new Date(validUntil)
    if (checkDate > until) {
      return false
    }
  }
  return true
}
